File: src/registry.ts
// ─── Enums ───────────────────────────────────────────────────

export type AgentId = "claude" | "opencode" | "codex" | "antigravity" | "copilot" | "droid";
export type ToolId = "rtk" | "caveman" | "codegraph" | "context-mode" | "ponytail" | "principles";
export type Channel = "github" | "npm" | "skill";

// ─── Info types ──────────────────────────────────────────────

export interface AgentInfo {
  id: AgentId;
  label: string;
  homepage: string;
  cliBin: string;
}

export interface ToolInfo {
  id: ToolId;
  label: string;
  homepage: string;
  channel: Channel;
  minNodeMajor: number;
  notTrackable?: boolean;
  instructionOnly?: boolean;
}

export interface Detection {
  installed: boolean;
  source: string;
}

export interface RunOpts {
  dryRun: boolean;
  upgrade: boolean;
  verbose: boolean;
  yes: boolean;
}

// ─── Static registries ──────────────────────────────────────

export const ALL_AGENTS: AgentInfo[] = [
  {
    id: "claude",
    label: "Claude Code",
    homepage: "https://github.com/anthropics/claude-code",
    cliBin: "claude",
  },
  {
    id: "opencode",
    label: "OpenCode",
    homepage: "https://github.com/anomalyco/opencode",
    cliBin: "opencode",
  },
  {
    id: "codex",
    label: "Codex",
    homepage: "https://github.com/openai/codex",
    cliBin: "codex",
  },
  {
    id: "antigravity",
    label: "Antigravity",
    homepage: "https://antigravity.google",
    cliBin: "agy",
  },
  {
    id: "copilot",
    label: "GitHub Copilot",
    homepage: "https://github.com/github/copilot-cli",
    cliBin: "copilot",
  },
  {
    id: "droid",
    label: "Factory Droid",
    homepage: "https://factory.ai",
    cliBin: "droid",
  },
];

export const ALL_TOOLS: ToolInfo[] = [
  {
    id: "rtk",
    label: "RTK",
    homepage: "https://github.com/rtk-ai/rtk",
    channel: "github",
    minNodeMajor: 0,
  },
  {
    id: "caveman",
    label: "Caveman",
    homepage: "https://github.com/JuliusBrussee/caveman",
    channel: "skill",
    minNodeMajor: 0,
  },
  {
    id: "codegraph",
    label: "CodeGraph",
    homepage: "https://github.com/colbymchenry/codegraph",
    channel: "npm",
    minNodeMajor: 18,
  },
  {
    id: "context-mode",
    label: "Context-Mode",
    homepage: "https://github.com/mksglu/context-mode",
    channel: "npm",
    minNodeMajor: 22,
  },
  {
    id: "ponytail",
    label: "Ponytail",
    homepage: "https://github.com/DietrichGebert/ponytail",
    channel: "npm",
    minNodeMajor: 0,
  },
  {
    id: "principles",
    label: "Principles",
    homepage: "https://github.com/multica-ai/andrej-karpathy-skills",
    channel: "skill",
    minNodeMajor: 0,
    notTrackable: true,
    instructionOnly: true,
  },
];

// ─── Lookups ─────────────────────────────────────────────────

export function agentInfo(id: AgentId): AgentInfo {
  return ALL_AGENTS.find((a) => a.id === id)!;
}

export function toolInfo(id: ToolId): ToolInfo {
  return ALL_TOOLS.find((t) => t.id === id)!;
}

export function parseAgentId(s: string): AgentId | null {
  const lower = s.toLowerCase().trim();
  if (["claude", "opencode", "codex", "antigravity", "copilot", "droid"].includes(lower)) {
    return lower as AgentId;
  }
  return null;
}

export function parseToolId(s: string): ToolId | null {
  const lower = s.toLowerCase().trim();
  const map: Record<string, ToolId> = {
    rtk: "rtk",
    caveman: "caveman",
    codegraph: "codegraph",
    "context-mode": "context-mode",
    contextmode: "context-mode",
    ponytail: "ponytail",
    principles: "principles",
    "karpathy-skills": "principles",
    karpathy: "principles",
    karpathyskills: "principles",
  };
  return map[lower] ?? null;
}

// ─── Dispatch functions ─────────────────────────────────────

import * as antigravity from "./agents/antigravity.js";
import * as claude from "./agents/claude.js";
import * as codex from "./agents/codex.js";
import * as copilot from "./agents/copilot.js";
import * as droid from "./agents/droid.js";
import * as opencode from "./agents/opencode.js";
import * as cavemanTool from "./tools/caveman.js";
import * as codegraphTool from "./tools/codegraph.js";
import * as contextModeTool from "./tools/context-mode.js";
import * as ponytailTool from "./tools/ponytail.js";
import * as principlesTool from "./tools/principles.js";
import * as rtkTool from "./tools/rtk.js";
import type { HealthStatus, RepairResult } from "./util/health.js";
import { getCachedLatest, getStaleFallback, setCachedLatest } from "./util/versioncache.js";

const agentModules = { claude, opencode, codex, antigravity, copilot, droid };
const toolModules = {
  rtk: rtkTool,
  caveman: cavemanTool,
  codegraph: codegraphTool,
  "context-mode": contextModeTool,
  ponytail: ponytailTool,
  principles: principlesTool,
} as const;

export function detectAgent(id: AgentId): Detection {
  return agentModules[id].detect();
}

export async function installTool(id: ToolId, opts: RunOpts): Promise<boolean> {
  return toolModules[id].install(opts);
}

export async function wireTool(agent: AgentId, tool: ToolId, opts: RunOpts): Promise<boolean> {
  return agentModules[agent].wire(tool, opts);
}

export async function unwireTool(agent: AgentId, tool: ToolId, opts: RunOpts): Promise<boolean> {
  return agentModules[agent].unwire(tool, opts);
}

export function verifyTool(agent: AgentId, tool: ToolId): boolean | null {
  return agentModules[agent].verify(tool);
}

export function toolInstalledVersion(id: ToolId): string | null {
  return toolModules[id].installedVersion();
}

export async function toolLatestVersion(id: ToolId): Promise<string | null> {
  // Check cache first (6h TTL)
  const cached = getCachedLatest(id);
  if (cached.hit) return cached.version;

  // Fetch from network
  const version = await toolModules[id].latestVersion();
  if (version) {
    setCachedLatest(id, version);
    return version;
  }

  // Fetch failed — fall back to stale cache if available
  return getStaleFallback(id);
}

export function toolHealthCheck(id: ToolId): HealthStatus {
  return toolModules[id].healthCheck();
}

export async function toolRepair(id: ToolId, opts: RunOpts): Promise<RepairResult> {
  return toolModules[id].repair(opts);
}

File: rust/src/registry.rs
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum AgentId {
    Claude,
    Opencode,
    Codex,
    Antigravity,
    Copilot,
    Droid,
    Devin,
    Warp,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum ToolId {
    Rtk,
    Caveman,
    Codegraph,
    ContextMode,
    Ponytail,
    Principles,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Channel {
    Github,
    Npm,
    Skill,
}

#[derive(Debug, Clone, Copy)]
pub struct AgentInfo {
    pub id: AgentId,
    pub label: &'static str,
    pub homepage: &'static str,
    pub cli_bin: &'static str,
}

#[derive(Debug, Clone, Copy)]
pub struct ToolInfo {
    pub id: ToolId,
    pub label: &'static str,
    pub homepage: &'static str,
    pub channel: Channel,
    pub min_node_major: u32,
    pub not_trackable: bool,
    pub instruction_only: bool,
}

#[derive(Debug, Clone, Copy, Default)]
pub struct RunOpts {
    pub dry_run: bool,
    pub upgrade: bool,
    pub verbose: bool,
    pub yes: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct Detection {
    pub installed: bool,
    pub source: String,
}

pub const ALL_AGENTS: &[AgentInfo] = &[
    AgentInfo {
        id: AgentId::Claude,
        label: "Claude Code",
        homepage: "https://github.com/anthropics/claude-code",
        cli_bin: "claude",
    },
    AgentInfo {
        id: AgentId::Opencode,
        label: "OpenCode",
        homepage: "https://github.com/anomalyco/opencode",
        cli_bin: "opencode",
    },
    AgentInfo {
        id: AgentId::Codex,
        label: "Codex",
        homepage: "https://github.com/openai/codex",
        cli_bin: "codex",
    },
    AgentInfo {
        id: AgentId::Antigravity,
        label: "Antigravity",
        homepage: "https://antigravity.google",
        cli_bin: "agy",
    },
    AgentInfo {
        id: AgentId::Copilot,
        label: "GitHub Copilot",
        homepage: "https://github.com/github/copilot-cli",
        cli_bin: "copilot",
    },
    AgentInfo {
        id: AgentId::Droid,
        label: "Factory Droid",
        homepage: "https://factory.ai",
        cli_bin: "droid",
    },
    AgentInfo {
        id: AgentId::Devin,
        label: "Devin / Cascade",
        homepage: "https://devin.ai",
        cli_bin: "devin",
    },
    AgentInfo {
        id: AgentId::Warp,
        label: "Warp / Oz",
        homepage: "https://warp.dev",
        cli_bin: "warp",
    },
];

pub const ALL_TOOLS: &[ToolInfo] = &[
    ToolInfo {
        id: ToolId::Rtk,
        label: "RTK",
        homepage: "https://github.com/rtk-ai/rtk",
        channel: Channel::Github,
        min_node_major: 0,
        not_trackable: false,
        instruction_only: false,
    },
    ToolInfo {
        id: ToolId::Caveman,
        label: "Caveman",
        homepage: "https://github.com/JuliusBrussee/caveman",
        channel: Channel::Skill,
        min_node_major: 0,
        not_trackable: false,
        instruction_only: false,
    },
    ToolInfo {
        id: ToolId::Codegraph,
        label: "CodeGraph",
        homepage: "https://github.com/colbymchenry/codegraph",
        channel: Channel::Npm,
        min_node_major: 18,
        not_trackable: false,
        instruction_only: false,
    },
    ToolInfo {
        id: ToolId::ContextMode,
        label: "Context-Mode",
        homepage: "https://github.com/mksglu/context-mode",
        channel: Channel::Npm,
        min_node_major: 22,
        not_trackable: false,
        instruction_only: false,
    },
    ToolInfo {
        id: ToolId::Ponytail,
        label: "Ponytail",
        homepage: "https://github.com/DietrichGebert/ponytail",
        channel: Channel::Npm,
        min_node_major: 0,
        not_trackable: false,
        instruction_only: false,
    },
    ToolInfo {
        id: ToolId::Principles,
        label: "Principles",
        homepage: "https://github.com/multica-ai/andrej-karpathy-skills",
        channel: Channel::Skill,
        min_node_major: 0,
        not_trackable: true,
        instruction_only: true,
    },
];

pub fn agent_info(id: AgentId) -> &'static AgentInfo {
    ALL_AGENTS
        .iter()
        .find(|a| a.id == id)
        .expect("valid agent id")
}

pub fn tool_info(id: ToolId) -> &'static ToolInfo {
    ALL_TOOLS
        .iter()
        .find(|t| t.id == id)
        .expect("valid tool id")
}

pub fn parse_agent_id(s: &str) -> Option<AgentId> {
    match s.to_lowercase().trim() {
        "claude" => Some(AgentId::Claude),
        "opencode" => Some(AgentId::Opencode),
        "codex" => Some(AgentId::Codex),
        "antigravity" => Some(AgentId::Antigravity),
        "copilot" => Some(AgentId::Copilot),
        "droid" => Some(AgentId::Droid),
        "devin" => Some(AgentId::Devin),
        "cascade" => Some(AgentId::Devin),
        "warp" => Some(AgentId::Warp),
        "oz" => Some(AgentId::Warp),
        _ => None,
    }
}

pub fn parse_tool_id(s: &str) -> Option<ToolId> {
    match s.to_lowercase().trim() {
        "rtk" => Some(ToolId::Rtk),
        "caveman" => Some(ToolId::Caveman),
        "codegraph" => Some(ToolId::Codegraph),
        "context-mode" | "contextmode" => Some(ToolId::ContextMode),
        "ponytail" => Some(ToolId::Ponytail),
        "principles" | "karpathy-skills" | "karpathy" | "karpathyskills" => {
            Some(ToolId::Principles)
        }
        _ => None,
    }
}
